Allow uprobes to work across filesystem namespace boundaries

ajor opened the pull request:
This is a rebased version of #2662, with fixes to pass the tests on CI.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

jordalgo commented on src/utils.cpp:
```diff
@@ -1453,7 +1453,22 @@ std::map<uintptr_t, elf_symbol, std::greater<>> get_symbol_table_for_elf(
+  auto program_abs = std_filesystem::canonical(program, ec);
+  if (ec)
+  {
+    // std::filesystem::canonical will fail if the we are attaching to
```
```suggestion
    // std::filesystem::canonical will fail if we are attaching to
```

jordalgo commented on src/utils.cpp:
```diff
@@ -1453,7 +1453,22 @@ std::map<uintptr_t, elf_symbol, std::greater<>> get_symbol_table_for_elf(
+    // used to attach to all running processes for a given binary,
+    // and the above uprobe is targetting a specific process.
+    // So if this happens, just return no pids.
+    // The probe will still attach directly to the targetted process.
```
```suggestion
    // The probe will still attach directly to the targeted process.
```

jordalgo commented on tests/CMakeLists.txt:
```diff
@@ -125,13 +125,17 @@ endif(NOT STATIC_LINKING)
 file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/testprogs/)
 file(GLOB testprogs testprogs/*.c)
 set(compiled_testprogs "")
+
```
extra break

jordalgo commented on tests/testprogs/mountns_pivot_wrapper.c:
```diff
@@ -0,0 +1,88 @@
+usage: mountns_pivot_wrapper some_testprog
+
+This wrapper is very similar to mountns_wrapper, but it pivots the root
+before executing the programk. Docker and containerd use pivot_root to
```
```suggestion
before executing the program. Docker and containerd use pivot_root to
```

jordalgo commented on tests/testprogs/uprobe_test_static.c:
```diff
@@ -0,0 +1,18 @@
+#include <unistd.h>
+
+int function1(int *n)
```
I guess we don't need to add the noinline directive here yea?

viktormalik commented on CHANGELOG.md:
```diff
@@ -107,7 +109,6 @@ and this project adheres to
 - Improve detection of unknown typedefs in ClangParser
   - [#2763](https://github.com/iovisor/bpftrace/pull/2763)
 
-
```
Not related to this change, remove?

viktormalik reviewed (approved):
I'm not much familiar with the pivot root, but otherwise looks good to me. Thanks for reviving this!

ajor commented:
The old tests did not compile on my machine or on the CI. Updated the test setup to work with dynamically linked executables and removed the static linking stuff.

jordalgo commented on tests/testprogs/mountns_pivot_wrapper.c:
```diff
@@ -0,0 +1,127 @@
+  char oldroot[PATH_MAX];
+
+  if (argc != 2)
+    errExit("Must specify test program as only argument.");
```
nit
```suggestion
    errExit("mountns_pivot_wrapper only accepts a single argument.");
```

viktormalik commented on tests/testprogs/mountns_pivot_wrapper.c:
```diff
@@ -0,0 +1,127 @@
+realpath of the executable will report paths in the mounted namespace
+rather than in the host namespace.
+
+Any binary for use with this wrapper should be compiled as static as it
+will not have access to shared libraries.
```
This is not the case anymore, is is?

ajor replied:
Nope, fixed
